clang/lib/Headers/hlsl/hlsl_intrinsic_helpers.h

```diff
@@ -71,6 +71,23 @@ constexpr vector<T, L> reflect_vec_impl(vector<T, L> I, vector<T, L> N) {
 #endif
 }
 
+template <typename T> constexpr T refract_impl(T I, T N, T Eta) {
+  T K = 1 - Eta * Eta * (1 - (N * I * N * I));
+  T Result = (Eta * I - (Eta * N * I + sqrt(K)) * N);
+  return select<T>(K < 0, static_cast<T>(0), Result);
+}
+
+template <typename T, int L>
+constexpr vector<T, L> refract_vec_impl(vector<T, L> I, vector<T, L> N, T Eta) {
+#if (__has_builtin(__builtin_spirv_refract))
+  return __builtin_spirv_refract(I, N, Eta);
+#else
+  vector<T, L> K = 1 - Eta * Eta * (1 - dot(N, I) * dot(N, I));
```

Comment:
we should really be storing `dot(N, I)` somewhere so that we aren't repeating the computation 3 times. That should simplify the codegen for the -O0 cases.